### PR TITLE
Add idle time support

### LINKED ISSUE #179: create Idle Time logger for linux

### LINKED ISSUE #180: create Idle Time logger for macos

### PR DESCRIPTION
**Pull request by rundong08**
Fixes #179 
Fixes #180

### CODE REVIEW

File: pal_event_server/lib/src/loggers/app_usage/idle_time.dart [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import 'package:logging/logging.dart';
+
+final _logger = Logger('IdleTime');
+
+RegExp pattern = RegExp(r'"HIDIdleTime" = (\d+)');
+
+Future<double> getIdleSecondsMacOS() async {
+  final result = await Process.run('ioreg', ['-rc' 'IOHIDSystem']);
+  final match = pattern.firstMatch(result.stdout);
+  if (match == null || match[1] == null) {
```
Comment (BobEvans) [inline pal_event_server/lib/src/loggers/app_usage/idle_time.dart:26]: Do we need to check for an out of bounds access on the match[1] prior to accessing it?

File: pal_event_server/lib/src/loggers/app_usage/linux/linux_helper.dart [hunk 2]
```diff
@@ -34,9 +36,13 @@ void linuxAppLoggerIsolate(SendPort sendPort) {
...
-          // Gets the active window name
-          Process.run(xprop.command, xprop.getAppArgs(windowId)).then((result) {
-            final currWindow = result.stdout;
+          // Gets the active window name and idle state
+          Future.wait([
+            Process.run(xprop.command, xprop.getAppArgs(windowId)),
+            isIdle()
+          ]).then((result) {
+            final currWindow = (result[0] as ProcessResult).stdout;
+            final idleState = result[1];
```
Comment (BobEvans) [inline pal_event_server/lib/src/loggers/app_usage/linux/linux_helper.dart:45]: what happens if the process crashes? Can result[1] ever cause an out of bounds access problem?

File: pal_event_server/lib/src/loggers/app_usage/macos/macos_helper.dart [hunk 3]
```diff
@@ -18,39 +18,53 @@ import 'dart:async';
...
   Timer.periodic(queryInterval, (Timer _) {
     try {
-      Process.run(apple_script.command, apple_script.scriptArgs).then((result) {
-        final currWindow = result.stdout.trim();
+      Future.wait([
+        Process.run(apple_script.command, apple_script.scriptArgs),
+        isIdle()
+      ]).then((result) {
+        final currWindow = (result[0] as ProcessResult).stdout.trim();
+        final idleState = result[1];
```
Comment (BobEvans) [inline pal_event_server/lib/src/loggers/app_usage/macos/macos_helper.dart:40]: is it possible to get an out of bounds error here?

**Comment by BobEvans** [comment]
Thanks! Looks great! Just a few small questions.

File: pal_event_server/lib/src/loggers/app_usage/linux/linux_helper.dart (hunk 2 above)
Comment (rundong08) [reply pal_event_server/lib/src/loggers/app_usage/linux/linux_helper.dart:45]: If the process crashes, ProcessResult will still be generated with an non-zero exit code, the stdout will probably be empty or garbage but those were handled. If one of the `Process.run` throws an exception then the `then` body will not be executed and the error will be caught by the `catchError` below.

File: pal_event_server/lib/src/loggers/app_usage/macos/macos_helper.dart (hunk 3 above)
Comment (rundong08) [reply pal_event_server/lib/src/loggers/app_usage/macos/macos_helper.dart:40]: Same as above

File: pal_event_server/lib/src/loggers/app_usage/linux/linux_helper.dart (hunk 2 above)
Comment (BobEvans) [reply pal_event_server/lib/src/loggers/app_usage/linux/linux_helper.dart:45]: Thanks

File: pal_event_server/lib/src/loggers/app_usage/macos/macos_helper.dart (hunk 3 above)
Comment (BobEvans) [reply pal_event_server/lib/src/loggers/app_usage/macos/macos_helper.dart:40]: Thanks

**Pull request merged**